cpp/mcap/include/mcap/errors.hpp

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include <string>
+#include <system_error>
+
+namespace mcap {
+
+enum class ErrorCode {
+  Success = 0,
+  NotOpen = 1,
+  InvalidChannelId = 2,
+};
+
+}  // namespace mcap
+
+namespace std {
+// Register mcap::ErrorCode with the standard error code system
+template <>
+struct is_error_code_enum<mcap::ErrorCode> : true_type {};
+}  // namespace std
+
+namespace mcap {
+
+namespace detail {
+
+// Define a custom error code category derived from std::error_category
+class McapErrorCategory : public std::error_category {
+public:
+  virtual const char* name() const noexcept override final {
+    return "McapError";
+  }
+
+  virtual std::string message(int c) const override final {
+    switch (static_cast<ErrorCode>(c)) {
+      case ErrorCode::Success:
+        return "success";
+      case ErrorCode::NotOpen:
+        return "not open";
+      case ErrorCode::InvalidChannelId:
+        return "invalid channel id";
+      default:
+        return "unknown";
+    }
+  }
+
+  virtual std::error_condition default_error_condition(int c) const noexcept override final {
+    switch (static_cast<ErrorCode>(c)) {
+      case ErrorCode::NotOpen:
+        return make_error_condition(std::errc::bad_file_descriptor);
+      case ErrorCode::InvalidChannelId:
+        return make_error_condition(std::errc::invalid_argument);
```

Comment:
not very familiar with the `default_error_condition` API, but why would we want to translate to posix error codes when we already have a custom one that exactly describes the error?

Reply:
This is an optional override in the API for users that want to compare your custom error codes against posix error codes. Doing more specific comparisons is always encouraged, but I guess the thinking is if this error happens several levels deep in your codebase and it bubbles up to a generic error handler that has different handling for filesystem or I/O errors vs application logic errors (like `invalid_argument`).

To answer your question, "we" would probably never do this, but it's a common nicety for library users that may do this for whatever reason.